python/lsst/ts/observatory/control/base_group.py

```diff
@@ -0,0 +1,303 @@
+# This file is part of ts_observatory_control.
+#
+# Developed for the LSST Telescope and Site Systems.
+# This product includes software developed by the LSST Project
+# (https://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+
+import types
+import asyncio
+import logging
+
+from lsst.ts import salobj
+
+__all__ = ["BaseGroup"]
+
+
+class BaseGroup:
```

Comment:
What about `RemoteCollection` ?

Reply:
Sure.

A thought that might inform your choice. I suggested making the remotes available as attributes named `rem_<SAL component name>` -- i.e. adding the "rem_" prefix to avoid any danger of conflicts.

A similar option that you might consider (and one I'd probably use in salobj if I was starting over) is to hold the remotes in a `types.SimpleNamespace` named "rem". (In the case if salobj I'd have 3 of them: "cmd" for command topics, "evt" for event topics and "tel" for  telemetry topics. Then access would be like this:

`mywhatever.rem.ATPtg` instead of `mywhatever.rem_ATPtg`.

Advantages:
* No need for setattr to create the collection, so the code is simpler.
* Easy to iterate over all contained remotes/events/etc. without having to keep your own dicts/lists/whatever.

Disadvantages:
* dir(foo) shows the simple namespace object(s), but not what is held in them.

Anyway, if you go this route then the actual "remote collection" is a SimpleNamespace and your class holds that and has some convenience methods for operating with them. Thus the name RemoteCollection might not be quite so appropriate.

Sorry if this is anti-helpful.

Reply:
I like the suggestion but I wonder if it would be really necessary. I understand that salobj has three distinct types for "thing" (commands, events and telemetry) to distinguish, whereas this class only has the remotes. 

I would also have to implement some sort of deprecation for the current way to access the remotes, or risk having users loosing a lot of notebooks functionality. I'll have to think a bit more about this.

Reply:
I consider it very important because it eliminates the risk of a namespace collision.

It really isn't related to having 1 or N categories of objects. It is strictly driven by adding attributes that can have nearly any name. These should be added to a special object so their names they can never collide with your method names. This sort of problem could easily crop up later -- when somebody expands the code or adds a new SAL component whose name conflicts with an existing method name. Best to fix it before it bites us.

Reply:
Ok, that is reasonable... I'll have to add a deprecation period though.